File: smash-ranks-android/app/src/main/java/com/garpr/android/models/AbsPlayer.java
package com.garpr.android.models;

import android.os.Parcel;
import android.os.Parcelable;

import com.google.gson.JsonDeserializationContext;
import com.google.gson.JsonDeserializer;
import com.google.gson.JsonElement;
import com.google.gson.JsonObject;
import com.google.gson.JsonParseException;
import com.google.gson.JsonSerializationContext;
import com.google.gson.JsonSerializer;
import com.google.gson.annotations.SerializedName;

import java.lang.reflect.Type;
import java.util.Comparator;

public abstract class AbsPlayer implements Parcelable {

    @SerializedName("id")
    protected String mId;

    @SerializedName("name")
    protected String mName;


    @Override
    public boolean equals(final Object obj) {
        return obj instanceof AbsPlayer && mId.equals(((AbsPlayer) obj).getId());
    }

    public String getId() {
        return mId;
    }

    public abstract Kind getKind();

    public String getName() {
        return mName;
    }

    @Override
    public int hashCode() {
        return mId.hashCode();
    }

    @Override
    public String toString() {
        return getName();
    }

    @Override
    public int describeContents() {
        return 0;
    }

    protected void readFromParcel(final Parcel source) {
        mId = source.readString();
        mName = source.readString();
    }

    @Override
    public void writeToParcel(final Parcel dest, final int flags) {
        dest.writeString(mId);
        dest.writeString(mName);
    }

    public static final Comparator<AbsPlayer> ALPHABETICAL_ORDER = new Comparator<AbsPlayer>() {
        @Override
        public int compare(final AbsPlayer o1, final AbsPlayer o2) {
            return o1.getName().compareToIgnoreCase(o2.getName());
        }
    };

    public static final JsonDeserializer<AbsPlayer> JSON_DESERIALIZER = new JsonDeserializer<AbsPlayer>() {
        @Override
        public AbsPlayer deserialize(final JsonElement json, final Type typeOfT,
                final JsonDeserializationContext context) throws JsonParseException {
            if (json == null || json.isJsonNull()) {
                return null;
            }

            final JsonObject jsonObject = json.getAsJsonObject();

            if (jsonObject.has("region")) {
                return context.deserialize(json, FavoritePlayer.class);
            } else if (jsonObject.has("aliases") || jsonObject.has("regions") || jsonObject.has("ratings")) {
                return context.deserialize(json, FullPlayer.class);
            } else {
                return context.deserialize(json, LitePlayer.class);
            }
        }
    };

    public static final JsonSerializer<AbsPlayer> JSON_SERIALIZER = new JsonSerializer<AbsPlayer>() {
        @Override
        public JsonElement serialize(final AbsPlayer src, final Type typeOfSrc,
                final JsonSerializationContext context) {
            if (src == null) {
                return null;
            }

            switch (src.getKind()) {
                case FAVORITE:
                    return context.serialize(src, FavoritePlayer.class);

                case FULL:
                    return context.serialize(src, FullPlayer.class);

                case LITE:
                    return context.serialize(src, LitePlayer.class);
            }

            throw new RuntimeException("unknown kind: \"" + src.getKind() + "\"");
        }
    };


    public enum Kind implements Parcelable {
        @SerializedName("favorite")
        FAVORITE,

        @SerializedName("full")
        FULL,

        @SerializedName("lite")
        LITE;

        @Override
        public int describeContents() {
            return 0;
        }

        @Override
        public void writeToParcel(final Parcel dest, final int flags) {
            dest.writeInt(ordinal());
        }

        public static final Creator<Kind> CREATOR = new Creator<Kind>() {
            @Override
            public Kind createFromParcel(final Parcel source) {
                return values()[source.readInt()];
            }

            @Override
            public Kind[] newArray(final int size) {
                return new Kind[size];
            }
        };
    }

}

File: smash-ranks-android/app/src/main/java/com/garpr/android/models/AbsPlayer.kt
package com.garpr.android.models

import android.os.Parcel
import android.os.Parcelable
import com.garpr.android.extensions.createParcel
import com.google.gson.JsonDeserializer
import com.google.gson.JsonSerializer
import com.google.gson.annotations.SerializedName
import java.util.Comparator

abstract class AbsPlayer(
        @SerializedName("id") val id: String,
        @SerializedName("name") val name: String
) : Parcelable {

    companion object {
        val ALPHABETICAL_ORDER = Comparator<AbsPlayer> { o1, o2 ->
            o1.name.compareTo(o2.name, ignoreCase = true)
        }

        val JSON_DESERIALIZER = JsonDeserializer<AbsPlayer> { json, typeOfT, context ->
            if (json == null || json.isJsonNull) {
                return@JsonDeserializer null
            }

            val jsonObject = json.asJsonObject

            if (jsonObject.has("region")) {
                context.deserialize<FavoritePlayer>(json, FavoritePlayer::class.java)
            } else if (jsonObject.has("aliases") ||
                    jsonObject.has("regions") ||
                    jsonObject.has("ratings")) {
                context.deserialize<FullPlayer>(json, FullPlayer::class.java)
            } else if (jsonObject.has("ranked") ||
                    jsonObject.has("rating") ||
                    jsonObject.has("previous_rank")) {
                context.deserialize<RankedPlayer>(json, RankedPlayer::class.java)
            } else {
                context.deserialize<LitePlayer>(json, LitePlayer::class.java)
            }
        }

        val JSON_SERIALIZER = JsonSerializer<AbsPlayer> { src, typeOfSrc, context ->
            when (src?.kind) {
                Kind.FAVORITE -> context.serialize(src, FavoritePlayer::class.java)
                Kind.FULL -> context.serialize(src, FullPlayer::class.java)
                Kind.LITE -> context.serialize(src, LitePlayer::class.java)
                Kind.RANKED -> context.serialize(src, RankedPlayer::class.java)
                else -> null
            }
        }
    }

    override fun equals(other: Any?): Boolean {
        return other is AbsPlayer && id.equals(other.id, ignoreCase = true)
    }

    override fun hashCode() = id.hashCode()

    abstract val kind: Kind

    override fun toString() = name

    override fun describeContents() = 0

    override fun writeToParcel(dest: Parcel, flags: Int) {
        dest.writeString(id)
        dest.writeString(name)
    }


    enum class Kind : Parcelable {
        @SerializedName("favorite")
        FAVORITE,

        @SerializedName("full")
        FULL,

        @SerializedName("lite")
        LITE,

        @SerializedName("ranked")
        RANKED;

        companion object {
            @JvmField
            val CREATOR = createParcel { values()[it.readInt()] }
        }

        override fun describeContents() = 0

        override fun writeToParcel(dest: Parcel, flags: Int) {
            dest.writeInt(ordinal)
        }
    }

}
